LFS-816: Handle re-login globally (continued)

IntegralProgrammer opened the pull request:
If these changes are acceptable, please merge the `LFS-816` branch into `dev` but do not close this Pull Request or delete the `LFS-816` branch as `fetchWithReLogin()` needs to be added to various other places. I have opened this pull request now so as to keep the `LFS-816` and `dev` branches from drifting too far apart and causing merge conflicts.

- Upgrades DeleteButton.jsx from using `fetch()` to using `fetchWithReLogin()`
  - To test:
    - Build the LFS-816 branch
    - Create a new _Demographics_ Form
    - Enter some data and save the Form
    - Open a new browser tab, visit `http://localhost:8080` and logout.
    - Return to the original browser tab
    - Click the _Trash Can_ icon to delete the Form
    - Confirm Form deletion
    - You should be prompted with a login dialog
    - After properly logging in, the Form should be deleted

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

Upgrade DeleteButton.jsx from fetch() to fetchWithReLogin()

marta- reviewed (changes requested):
There is code handling re-login "the old way" that needs to be removed. See functions handleError (line 72), handleDeleteButtonClicked (line 115), handleLogin (line 123) for example.

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

Remove obsolete invalid session handling in DeleteButton.jsx

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

When the fetchWithReLogin() login dialog is open, allow fetchWithReLogin() requests to build up in a queue so that they can be released upon login.

If it is not desired that a fetchWithReLogin() request be placed in a queue, call fetchWithReLogin with the noQueue argument set to true.

marta- commented on modules/login/src/main/frontend/src/login/loginDialogue.js:
```diff
@@ -23,16 +23,16 @@ import MainLoginComponent from './loginMainComponent';
 export const GlobalLoginContext = React.createContext();
 
-export function fetchWithReLogin(displayLoginCtx, url, fetchArgs) {
+export function fetchWithReLogin(displayLoginCtx, url, fetchArgs, noQueue) {
```
Negations in variable names should be avoided.One suggestion would be `discardOnFailure`, other suggestions are welcome.

sdumitriu commented on modules/login/src/main/frontend/src/login/loginDialogue.js:
```diff
@@ -23,16 +23,16 @@ import MainLoginComponent from './loginMainComponent';
-              displayLoginCtx.dialogOpen(fetchFunc);
-          } else if (response.ok) {
+              displayLoginCtx.dialogOpen(fetchFunc, noQueue);
+          } else if (response.ok || response.status == 404 || response.status == 409) {
```
Most of our code so far is used to handling error messages in the "resolved" case by checking for `response.ok`, so should this always resolve here instead of just for certain status codes?

IntegralProgrammer replied:
That is a good idea - in order to be consistent with JavaScript's `fetch()` API, I have reworked `fetchWithReLogin()` in commit `9cd68d2` so that `resolve()` is called regardless of status code (just like in `fetch()`) and `reject()` is only called if the underlying `fetch()` call generates an exception.

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

Even if requests are not supposed to be added to a queue, add one if the queue is empty - otherwise this type of request would always be dropped

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

Code clean-up / variable renaming

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -89,7 +89,7 @@ function UserDashboard(props) {
       .finally(() => setCreationLoading(false));
   }, [])
 
-  if (loading || globalLoginDisplay.getDialogOpenStatus()) {
```
`globalLoginDisplay` doesn't seem used anymore, can it be removed from this file?

IntegralProgrammer replied:
Yes, it can. I have addressed this in commit `abff3fa` :heavy_check_mark:

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -103,17 +103,19 @@ class Main extends React.Component {
-              success && this.setState({
-                loginDialogOpen: false,
-                loginHandler: undefined
+              success && this.state.loginDialogOpen && this.setState({
```
Why do you need to check `this.state.loginDialogOpen`? I think that just setting it to false more than one time should have no side effects.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -129,7 +131,17 @@ class Main extends React.Component {
+        <DialogueLoginContainer
+          isOpen={this.state.loginDialogOpen}
+          handleLogin={(success) => {
+            if (success) {
```
If `success` is false, we don't call the handlers, and we don't unset them. I didn't check how this affects the rest of the code, but are we sure this is what we want?

IntegralProgrammer replied:
Yes it is, if we always unset the handlers regardless of the value of `success`, and a login dialog appeared and the user entered incorrect credentials (thus setting `success` to `false`) the login handlers would be empty and the login dialog would not work anymore even with the correct credentials.

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

Code clean-up / Remove unnecessary GlobalLoginContext in UserDashboard.jsx

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

Always resolve any fetchWithReLogin() provided that the underlying fetch() call resolves and does not return a login-required state

sashaandjic commented:
re-tested

IntegralProgrammer committed:
LFS-816: Handle re-login globally (continued)

Remove unnecessary check of this.state.loginDialogOpen

sashaandjic commented:
tested the following:fetchWithReLogin() should be working with the following features:
User dashboard
Subjects page
Forms page
Any individual subject page (eg. a patient chart)
Any individual form (eg. a Demographics data form)
The delete button (trash can) anywhere (eg. Patient, Form)
